src/dune_engine/source_tree.ml

```diff
@@ -384,9 +391,24 @@ module Dir0 = struct
         Path.Source.Set.add acc (Path.Source.relative t.path s))
 end
 
-let ancestor_vcs = Fdecl.create Dyn.Encoder.opaque
-
-let init ~ancestor_vcs:m = Fdecl.set ancestor_vcs m
+let ancestor_vcs =
+  Memo.lazy_ (fun () ->
+      if Config.inside_dune then
+        Memo.Build.return None
+      else
+        let rec loop dir =
+          if Fpath.is_root dir then
+            None
+          else
+            let dir = Filename.dirname dir in
+            match
+              Sys.readdir dir |> Array.to_list |> String.Set.of_list
+              |> Vcs.Kind.of_dir_contents
```

Comment:
A bit of a shame to have to go through 2 pointless conversions just to check for the existence of 2 filenames.

Reply:
Agreed